### PR TITLE
[enhancement](multi-catalog)Persist external catalog related metadata.

### PR DESCRIPTION
**Pull request by Jibing-Li**
# Proposed changes

Issue Number: close #xxx

## Problem summary

Persist external catalog/db/table, including the columns of external tables. After this change, external objects could have their own uniq ID through their lifetime, this is required for the statistic information collection.
Fix the bug that Observer trying to get the uniq ID.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/qe/MasterCatalogExecutor.java [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+import org.apache.doris.common.ClientPool;
+import org.apache.doris.thrift.FrontendService;
+import org.apache.doris.thrift.TInitExternalCtlMetaRequest;
+import org.apache.doris.thrift.TInitExternalCtlMetaResult;
+import org.apache.doris.thrift.TNetworkAddress;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+public class MasterCatalogExecutor {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/qe/MasterCatalogExecutor.java:29]: Add some comment for this class

File: gensrc/thrift/FrontendService.thrift [hunk 2]
```diff
@@ -678,6 +678,17 @@ struct TWaitingTxnStatusResult {
...
 }
 
+struct TInitExternalCtlMetaRequest {
+    1: optional string catalogName
+    2: optional string dbName
+    3: optional string tableName
+}
+
+struct TInitExternalCtlMetaResult {
+    1: required i64 maxJournalId;
```
Comment (morningman) [inline gensrc/thrift/FrontendService.thrift:688]:
```suggestion
    1: optional i64 maxJournalId;
```

File: fe/fe-core/src/main/java/org/apache/doris/datasource/HMSExternalCatalog.java [hunk 3]
```diff
@@ -105,10 +106,24 @@ private void init() {
...
+                try {
+                    remoteExecutor.forward(name, null, null);
+                } catch (Exception e) {
+                    LOG.warn("Failed to forward init operation to master. {}", e.getMessage());
+                }
+                return;
+            }
             init();
             initialized = true;
+            ExternalObjectInitLog log = new ExternalObjectInitLog();
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/HMSExternalCatalog.java:124]: Missing database info?

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 40.2 seconds
 load time: 560 seconds
 storage size: 17154655374 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221028055207_clickbench_pr_35170.html

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 38.38 seconds
 load time: 559 seconds
 storage size: 17154699320 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221028063708_clickbench_pr_35189.html

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 39.18 seconds
 load time: 566 seconds
 storage size: 17154644841 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221028134451_clickbench_pr_35465.html

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 38.83 seconds
 load time: 554 seconds
 storage size: 17154644878 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221028161416_clickbench_pr_35509.html

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 39.06 seconds
 load time: 560 seconds
 storage size: 17154644851 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221028170437_clickbench_pr_35529.html

File: fe/fe-core/src/main/java/org/apache/doris/datasource/EsExternalCatalog.java [hunk 4]
```diff
@@ -138,28 +136,67 @@ private void validate(Map<String, String> properties) throws DdlException {
...
+            InitCatalogLog initCatalogLog = new InitCatalogLog();
+            init(initCatalogLog);
             initialized = true;
+            Env.getCurrentEnv().getEditLog().logInitCatalog(initCatalogLog);
+        }
+    }
+
+    @Override
+    public void addDb(ExternalDatabase db) {
+        if (!(db instanceof HMSExternalDatabase)) {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/EsExternalCatalog.java:172]:
```suggestion
        if (!(db instanceof EsExternalDatabase)) {
```

File: fe/fe-core/src/main/java/org/apache/doris/catalog/external/EsExternalDatabase.java [hunk 5]
```diff
@@ -60,13 +60,24 @@ public EsExternalDatabase(ExternalCatalog extCatalog, long id, String name) {
         super(extCatalog, id, name);
     }
 
-    public void addTable(ExternalTable table) {
-        if (!(table instanceof EsExternalTable)) {
-            LOG.warn("Table {} is not EsExternalTable.", table.getName());
-            return;
+    public synchronized void replayInitDb(InitDatabaseLog log, ExternalCatalog catalog) {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/external/EsExternalDatabase.java:63]: Do we need `synchronized`?

File: fe/fe-core/src/main/java/org/apache/doris/catalog/external/EsExternalDatabase.java [hunk 6]
```diff
@@ -135,12 +145,14 @@ public Set<String> getTableNamesWithLock() {
     @Override
     public List<EsExternalTable> getTables() {
         makeSureInitialized();
+        MasterCatalogExecutor.waitJournalReplay(replayToJournalId);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/external/EsExternalDatabase.java:148]: Why not move `MasterCatalogExecutor.waitJournalReplay(replayToJournalId);` into `makeSureInitialized()`?

File: fe/fe-core/src/main/java/org/apache/doris/datasource/ExternalCatalog.java [hunk 7]
```diff
@@ -85,12 +93,20 @@ public abstract class ExternalCatalog implements CatalogIf<ExternalDatabase>, Wr
...
-    public abstract void addDb(ExternalDatabase db);
-
     public void setInitialized(boolean initialized) {
         this.initialized = initialized;
     }
 
+    protected void waitJournalReplay() {
+        if (!Env.getCurrentEnv().isMaster()) {
+            try {
+                ConnectContext.get().getEnv().getJournalObservable().waitOn(replayToJournalId, 3000);
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/datasource/ExternalCatalog.java:103]: I think this should be done in `MasterCatalogExecutor`?

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**